server/src/main/java/org/elasticsearch/index/codec/tsdb/es819/ES819TSDBDocValuesConsumer.java

```diff
@@ -138,89 +143,126 @@ private long[] writeField(FieldInfo field, TsdbDocValuesProducer valuesProducer,
         meta.writeLong(numValues);
         meta.writeInt(numDocsWithValue);
 
-        if (numValues > 0) {
-            // Special case for maxOrd of 1, signal -1 that no blocks will be written
-            meta.writeInt(maxOrd != 1 ? ES819TSDBDocValuesFormat.DIRECT_MONOTONIC_BLOCK_SHIFT : -1);
-            final ByteBuffersDataOutput indexOut = new ByteBuffersDataOutput();
-            final DirectMonotonicWriter indexWriter = DirectMonotonicWriter.getInstance(
-                meta,
-                new ByteBuffersIndexOutput(indexOut, "temp-dv-index", "temp-dv-index"),
-                1L + ((numValues - 1) >>> ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SHIFT),
-                ES819TSDBDocValuesFormat.DIRECT_MONOTONIC_BLOCK_SHIFT
-            );
-
-            final long valuesDataOffset = data.getFilePointer();
-            // Special case for maxOrd of 1, skip writing the blocks
-            if (maxOrd != 1) {
-                final long[] buffer = new long[ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE];
-                int bufferSize = 0;
-                final TSDBDocValuesEncoder encoder = new TSDBDocValuesEncoder(ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE);
-                values = valuesProducer.getSortedNumeric(field);
-                final int bitsPerOrd = maxOrd >= 0 ? PackedInts.bitsRequired(maxOrd - 1) : -1;
-                for (int doc = values.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = values.nextDoc()) {
-                    final int count = values.docValueCount();
-                    for (int i = 0; i < count; ++i) {
-                        buffer[bufferSize++] = values.nextValue();
-                        if (bufferSize == ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE) {
-                            indexWriter.add(data.getFilePointer() - valuesDataOffset);
-                            if (maxOrd >= 0) {
-                                encoder.encodeOrdinals(buffer, data, bitsPerOrd);
-                            } else {
-                                encoder.encode(buffer, data);
+        IndexOutput disiTempOutput = null;
+        String skipListTempFileName = null;
+        IndexedDISIBuilder docIdSetBuilder = null;
```

Comment:
Should we make IndexedDISIBuilder (or maybe Accumulator is a better name?) Closable, pass a Directory in its constructor, and pass the IndexOutput in the build (or flush) method to copy the temporary output to the data output? This way, we only need a single reference here, making the code more manageable.